openwillis/measures/video/preprocess_video.py

```diff
@@ -0,0 +1,815 @@
+
+    return face_data_list 
+
+def extract_embed_faces_from_frame(
+    frame,
+    frame_idx,
+    rgb_back_ends,
+    detector_backend='mtcnn',
+    model_name='Facenet'
+):
+    """
+    ---------------------------------------------------------------------------------------------------
+
+    Extracts and embeds faces from a given frame using the specified backend and model.
+
+    Parameters:
+    ............
+    frame : np.ndarray
+        The input frame in BGR format.
+    frame_idx : int
+        The frame index.
+    detector_backend : str
+        The backend string specifying the face detection model to use.
+    model_name : str
+        Name of the model to use for embedding.
+
+    Returns:
+    ............
+    face_data_list : list
+        List of FaceData objects with embeddings.
+    ---------------------------------------------------------------------------------------------------
+    """
+    try:
+        face_data_list = extract_faces(
+            frame,
+            frame_idx,
+            rgb_back_ends,
+            back_end_str=detector_backend
+        )
+
+        face_data_list = prep_face_data_for_embed(face_data_list, frame)
+        face_data_list = embed_faces(face_data_list, model_name=model_name)
+    except Exception as e:
+        logger.error(f"Error extracting and embedding faces: {e}, frame_idx: {frame_idx}")
+        face_data_list = [FaceData()]
+    return face_data_list
+
+def load_facedata_from_video(
+    video_path,
+    rgb_back_ends,
+    detector_backend='mtcnn',
+    model_name='Facenet', 
+    n_frames=np.inf, 
+    capture_n_per_sec=3
+):
+    """
+    ---------------------------------------------------------------------------------------------------
+
+    Load video frames from a given video file, capturing only a specified
+    number of frames per second.
+
+    Parameters:
+    ............
+        video_path (str): Path to the video file.
+        n_frames (int, optional): Maximum number of frames to capture. Defaults to np.inf.
+        capture_n_per_sec (int, optional): Number of frames to capture per second. Defaults to 30.
+
+    Returns:
+    ............
+        tuple: A tuple containing the list of captured frames, the list of their corresponding frame indices,
+               and the frames per second (fps) of the video.
+    ---------------------------------------------------------------------------------------------------
+           
+    """
+    cap = cv2.VideoCapture(video_path)
+    if not cap.isOpened():
+        raise ValueError(f"Error opening video file: {video_path}")
+    
+    fps = cap.get(cv2.CAP_PROP_FPS)
+    if capture_n_per_sec > fps:
+        raise ValueError(f"Capture rate per second cannot exceed the video's frames per second, video fps: {fps}")
+
+    num_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+    n_skip = round(fps / capture_n_per_sec)
+    facedata_list = []
+
+    frame_index = 0
+    n_frames_capture = 0
```

Comment:
This looks like it was added for testing, if so it should be removed. If it actually is meaningful, we should remove the comments below instead.

Reply:
removed comments as I think this gives researchers a nice way to debug/try different detector backends /facemodels with out processing a full video. We could also delete though if you felt strongly to delete

Reply:
No that works for me